### ISSUE TITLE
GDC x86 32-bit support

### ISSUE THREAD

**Issue by p0nce** [AuburnSounds/intel-intrinsics#24]
- all x86 builtins are stable in GDC
- core.simd has more operands in GDC
- [x] can vector types be arbitrary length => YES
- assembly in GCC can be like `sprintf`, GCC will arrange operands (examples in druntime)
  Druntime's CPUID implementation is a good example, including how to preserve EBX
  this kind of assembly will be inline etc

- [x] First get basic support
- [x] ALL TESTS PASS on Travis (64-bit)
- [x] Introduce new enums GDC_with_SSE, GDC_with_MMX, GDC_with_SSE2, GDC_with_SSE3
- [x] replace GDC_x86 enum
- [x] use the generic path instead on GDC 32-bit, fix that path
- [x] ALL TESTS PASS on Travis (32-bit), 
- [x] Find a way to detectGDC_with_SSE, GDC_with_MMX, GDC_with_SSE2, GDC_with_SSE3 at compile-time so that to use builtins, or use an heuristic. For 32-bit GDC, disable those. (EDIT: later)

**Comment by p0nce** [comment]
- Interesting architectures in terms of GCC support: x86, RISC-V, ARM, AARCH64, PowerPC (rs6000), [sparc]
- Iain says most of his users are on x86

**Comment by p0nce** [comment]
- also the GCC Runtime Library Exceptions is very unclear to me that we'll be forced not to look at how GCC does it in `emmintrin.h` etc

**Comment by p0nce** [comment]
Iain updated https://explore.dgnu.org/ useful website

**Comment by p0nce** [comment]
GDC support increasingly useful for eg: https://forum.dlang.org/thread/izaufklyvmktnwsrmhci@forum.dlang.org?page=5

**Comment by p0nce** [comment]
- [ ] First get basic support
- [ ] get all SSE and SSE2 instructions to be optimized

**Comment by p0nce** [comment]
GDC from Travis and GDC from https://explore.dgnu.org disagree!
Blocked by https://travis-ci.community/t/could-someone-please-upgrade-the-gdc-travis-use/4533

**Comment by p0nce** [comment]
https://gdcproject.org/downloads is same version as Travis, Windows build there give the same errors, which is great.

**Comment by p0nce** [comment]
@baziotis 

**loadUnaligned**
For loadUnaligned it seems we just need to 

(A) check the template argument with is expression and select the right builtin, such as `__builtin_ia32_loadups` like you did in your SIMD module

OR

(B) use casts from/to static arrays to avoid the inherent alignedness of vector types

**Comment by p0nce** [comment]
As for **shufflevector**, I'm not sure we should use it for anything else than LDC. At first the idea was for it to be part of the public API of `intel-intrinsics`, but I think it needs to go in private internals instead.

**Comment by p0nce** [comment]
GDC extensions: https://wiki.dlang.org/Using_GDC#Extensions

**Comment by p0nce** [comment]
mini-SIMD module that implements some loadUnaligned versions for GCC https://github.com/dlang/druntime/pull/2687/files#diff-c2fcd73761ae6659ef91245ce1195b6dR70 It seems there is no other known way to have unaligned loads using GDC (we would need the aligned GCC attribute...)

**Comment by baziotis** [comment]
> @baziotis
> 
> **loadUnaligned**
> For loadUnaligned it seems we just need to
> 
> (A) check the template argument with is expression and select the right builtin, such as `__builtin_ia32_loadups` like you did in your SIMD module
> 
> OR
> 
> (B) use casts from/to static arrays to avoid the inherent alignedness of vector types

Sorry, I didn't get either of those. For (A), AFAIK GDC doesn't have available a `loadUnaligned` (as say LDC does, which is however different from `core.simd`). That means that we either:
a) Have to write our own. This makes sense to me because the current code uses a `loadUnaligned` everywhere. Then, depending on the template argument, we have to cast _back_ in order to return. I'm not sure whether you meant this with (B). If not, could you explain it a little more?
b) Have everywhere `static if` for the GDC version. Not very good option I think. And also, because there's no 1-1 mapping as with the rest of intel intrinsics, it doesn't seem a good choice.

So, if we do say the a), then remains a problem, which is, the aligned loads. In DMD, just say dereferencing a vector (e.g. https://github.com/AuburnSounds/intel-intrinsics/blob/master/source/inteli/emmintrin.d#L1546) results in aligned load. I guess to have it there the same is true for LDC. But for GDC I don't know.

**Comment by baziotis** [comment]
> As for **shufflevector**, I'm not sure we should use it for anything else than LDC. At first the idea was for it to be part of the public API of `intel-intrinsics`, but I think it needs to go in private internals instead.

The problem is that it is used in other intrinsics. I think it's lower priority than loads / stores. I'll definitely look at it when we finish those.

**Comment by p0nce** [comment]
I looked a bit more and `shufflevector` doesn't seem possible (fast) in the general case in GDC. It is in GCC, where it is available with the name `__builtin_shuffle`. But GDC doesn't have that.

**Comment by p0nce** [comment]
In GDC > 4.9.5, `vec.array[ind] = <scalar>;`  is a no-op! What should be used instead is `vec.ptr[ind] = <scalar>;`

**Comment by baziotis** [comment]
Do you have a code snippet for that? Oh that would be really surprising.
I can't reproduce, e.g.: https://d.godbolt.org/z/bg4KQW

**Comment by p0nce** [comment]
I can reproduce it very reliably on Debian with the GDC in gdcproject.com
See commits for full details of what is needed.

Example: => .array assignment work as of GDC 4.8.2 https://d.godbolt.org/z/wStswp
but doesn't in GDC 4.9.3 https://d.godbolt.org/z/MEifvc

That's not surprising at all, that sort of mess is why intel-intrinsics exist in the first place

**Comment by p0nce** [comment]
@baziotis the documentation is wrong then. What really matters is actual compilers.

**Comment by baziotis** [comment]
> That's not surprising at all, that sort of mess is why intel-intrinsics exist in the first place.

I agree. It's not so surprising if you think that `__vector` is internal compiler magic. In a normal struct member, I could never see this coming.

> @baziotis the documentation is wrong then. What really matters is actual compilers. Besides, i'm pretty sure the doc was telling otherwise months ago.

Indeed. Although it might be GDC that is wrong. We better report a bug although I'm not sure how GDC handles bug issues. [Bugzilla](https://bugzilla.gdcproject.org/) seems dead.

**Comment by p0nce** [comment]
Well we are well on our way to have GDC support, back to GDC built with DMD 2.068, with no overhead. It's less of a problem than compile-time SSE3 detection!

**Comment by p0nce** [comment]
https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html

**Comment by baziotis** [comment]
Be aware that inline asm should be avoided as much as possible because it's not visible to the optimizer. AFAIK, LLVM IR is ok exactly because it is IR.

**Comment by p0nce** [comment]
GDC can build unittest with 32-bit on Debian/Ubuntu with `sudo apt-get install gcc-multilib`
Else it won't link

**Comment by p0nce** [comment]
OK, GDC + 64-bit now pass (albeit some intrinsics are slow), but not all 32-bit unittests.

**Comment by baziotis** [comment]
Slow.. Really on x86_64?

**Comment by p0nce** [comment]
"slow" in the sense that sometimes the DMD path is taken when a GCC builtin would exist!

**Comment by baziotis** [comment]
Ah yes. You mean for the cases that we don't use a builtin because we can't version etc.. That versioning should be fixed soon.

**Comment by p0nce** [comment]
There are major problems with GDC + 32-bit x86, in the sense that -msse2 is seemingly somehow disabled by default and would probably require compile-time detection like SSE3. Basic stuff does't work, so I've postponed GDC + 32-bit x86 to further in the year.

**Comment by p0nce** [comment]
Course of actions:
- [ ] disable GDC_x86 enums for 32-bit
- [ ] use the generic path instead

**Comment by p0nce** [comment]
Available in both 32-bit and 64-bit as of v1.2.0
